io/io/src/RFile.cxx

```diff
@@ -178,51 +169,61 @@ static std::string ValidateAndNormalizePath(std::string &path)
    return "";
 }
 
-/////////////////////////////////////////////////////////////////////////////////////////////////
+static void EnsureFileOpenAndBinary(const TFile *tfile, std::string_view path)
+{
+   if (!tfile || tfile->IsZombie())
+      throw ROOT::RException(R__FAIL("failed to open file " + std::string(path) + " for reading"));
 
-RFile::RFile(std::unique_ptr<TFile> file) : fFile(std::move(file)) {}
+   if (tfile->IsRaw() || !tfile->IsBinary() || tfile->IsArchive())
+      throw ROOT::RException(R__FAIL("Opened file " + std::string(path) + " is not a ROOT binary file"));
+}
 
-RFile::~RFile() = default;
+/////////////////////////////////////////////////////////////////////////////////////////////////
+std::pair<std::string_view, std::string_view> ROOT::Experimental::DecomposePath(std::string_view path)
+{
+   auto lastSlashIdx = path.rfind('/');
+   if (lastSlashIdx == std::string_view::npos)
+      return {{}, path};
+
+   auto dirName = path.substr(0, lastSlashIdx + 1);
+   auto pathName = path.substr(lastSlashIdx + 1);
+   return {dirName, pathName};
+}
 
 std::unique_ptr<RFile> RFile::Open(std::string_view path)
 {
-   CheckExtension(path);
-
    TDirectory::TContext ctx(nullptr); // XXX: probably not thread safe?
    auto tfile = std::unique_ptr<TFile>(TFile::Open(std::string(path).c_str(), "READ_WITHOUT_GLOBALREGISTRATION"));
-   if (!tfile || tfile->IsZombie())
-      throw ROOT::RException(R__FAIL("failed to open file " + std::string(path) + " for reading"));
+   EnsureFileOpenAndBinary(tfile.get(), path);
 
    auto rfile = std::unique_ptr<RFile>(new RFile(std::move(tfile)));
    return rfile;
 }
 
 std::unique_ptr<RFile> RFile::Update(std::string_view path)
 {
-   CheckExtension(path);
-
    TDirectory::TContext ctx(nullptr); // XXX: probably not thread safe?
    auto tfile = std::unique_ptr<TFile>(TFile::Open(std::string(path).c_str(), "UPDATE_WITHOUT_GLOBALREGISTRATION"));
-   if (!tfile || tfile->IsZombie())
-      throw ROOT::RException(R__FAIL("failed to open file " + std::string(path) + " for updating"));
+   EnsureFileOpenAndBinary(tfile.get(), path);
 
    auto rfile = std::unique_ptr<RFile>(new RFile(std::move(tfile)));
    return rfile;
 }
 
 std::unique_ptr<RFile> RFile::Recreate(std::string_view path)
 {
-   CheckExtension(path);
-
    TDirectory::TContext ctx(nullptr); // XXX: probably not thread safe?
    auto tfile = std::unique_ptr<TFile>(TFile::Open(std::string(path).c_str(), "RECREATE_WITHOUT_GLOBALREGISTRATION"));
-   if (!tfile || tfile->IsZombie())
-      throw ROOT::RException(R__FAIL("failed to open file " + std::string(path) + " for writing"));
+   EnsureFileOpenAndBinary(tfile.get(), path);
 
    auto rfile = std::unique_ptr<RFile>(new RFile(std::move(tfile)));
    return rfile;
 }
 
+RFile::RFile(std::unique_ptr<TFile> file) : fFile(std::move(file)) {}
+
+RFile::~RFile() = default;
```

Comment:
Looking at this RFile destructor here also triggers my personal suggestion for employing rule of five.